### PR TITLE
Use `nil` checks for defaults when applicable

### PR DESCRIPTION
**Pull request by iwahbe**
Fixes https://github.com/pulumi/pulumi/issues/12106

# Description

We can see some passed values not respected when applying default values to interface types when the value passed to the interface is a plan value.

```go
type Args struct {
  Field pulumi.BoolPtrInput // default: true in the schema
}

(&Args{ Field: pulumi.BoolPtr(false) }).Defaults().Field // False
(&Args{ Field: pulumi.Bool(false) }).Defaults().Field // True
```

See https://go.dev/play/p/aiuA7Jzy0mH for a complete example.

We can fix this by switching the internal check of `Defaults()` to use a `== nil` check when an interface is being checked.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/go/gen.go
Comment (AaronFriel) [inline pkg/codegen/go/gen.go:1358]: Can we simplify this to remove the else case? 

Or rephrased, do we support defaults for non-optional fields?

Comment (AaronFriel) [reply pkg/codegen/go/gen.go:1358]: Grepping the PR branch, I only see 4 remaining calls to `isZero`:

* three in the `plain-and-default` test in `plain-and-default/go/foo/moduleResource.go`
* one in the `plain-object-defaults` test in `plain-object-defaults/go/example/pulumiTypes.go`

It seems safe to merge this if tests pass but I do wonder what exactly we're testing in those 4 cases. All of the fields involved are non-pointer, required fields in the schema. The codegen even looks nonsensical:

```
	if isZero(args.Plain_required_string) {
		args.Plain_required_string = "buzzer"
	}
```

I can almost see the argument for string or integer, but "isZero(args.Plain_required_boolean)" is nonsensical.

Comment (AaronFriel) [reply pkg/codegen/go/gen.go:1358]: For string and integer values, I wonder if we should instead be codegenning this for the zero case of required strings and booleans:

```
	if isZero(args.FieldName) {
		return nil, errors.New("invalid value for required argument 'ResourceNoAlias'")
	}
```

Or if we should be rejecting these programs during codegen as having invalid schemas. A required field with a default does not compute, unless the goal is to allow any value _except_ the zero value, forbidding empty strings, 0 and 0.0.

Comment (iwahbe) [reply pkg/codegen/go/gen.go:1358]: We have existing providers that set all combinations of `required`, `const` and `default` for the same field. We are keeping `isZero` and allowing these cases for backwards compatibility reasons. Fixing this is on the list of 4.0 changes.

**Review by jazzyfresh — APPROVED** [review]

**Review by abhinav — COMMENTED** [review]
Outside of the generated code, do we have a test exercising the actual generated behavior we want? Specifically, the behavior you have in the PR description against a generated test type?

**Comment by iwahbe** [comment]
> Outside of the generated code, do we have a test exercising the actual generated behavior we want? Specifically, the behavior you have in the PR description against a generated test type?

No, but there should be such I test. I'll add it.

**Review by abhinav — APPROVED** [review]
Made a couple minor tweaks to the test but LGTM. Would recommend squashing the commits before merging since.

**Comment by abhinav** [comment]
```
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -418,4 +418,10 @@
        	            	 
        	            	+        __props__.__dict__["optional_bool"] = None
        	            	         return ModuleResource(resource_name, opts=opts, __props__=__props__)
        	            	 
        	            	-
        	            	+    @property
        	            	+    @pulumi.getter
        	            	+    def optional_bool(self) -> pulumi.Output[Optional[bool]]:
        	            	+        return pulumi.get(self, "optional_bool")
        	            	+
        	            	+
        	Test:       	TestGeneratePackage/plain-and-default
        	Messages:   	pulumi_foobar/module_resource.py
    helpers.go:155: pulumi_foobar/module_resource.py did not agree
```

Might need to regenerate

**Comment by iwahbe** [comment]
bors retry

**Pull request merged**
